src/components/MapboxMap.tsx

```diff
@@ -0,0 +1,157 @@
+
+/**
+ * A function use to derive a result's coordinate.
+ *
+ * @public
+ */
+export type CoordinateGetter<T> = (result: Result<T>) => Coordinate | undefined;
+
+/**
+ * Coordinate use to represent the result's location on a map.
+ *
+ * @public
+ */
+export interface Coordinate {
+  /** The latitude of the location. */
+  latitude: number,
+  /** The longitude of the location. */
+  longitude: number
+}
+
+/**
+ * Props for the {@link MapboxMap} component.
+ *
+ * @public
+ */
+export interface MapboxMapProps<T> {
+  /** Mapbox access token. */
+  mapboxAccessToken: string,
+  /** Interface for customizing the {@link MapboxMap} map derived from Mapbox API's Map options. */
+  mapboxOptions?: Omit<MapboxOptions, 'container'>,
+  /** Custom Pin component to render for markers on the map. */
+  PinComponent?: PinComponent<T>,
+  /** A function to derive a result's coordinate for the corresponding marker's location on the map. */
+  getCoordinate?: CoordinateGetter<T>,
+  /** A function which is called when user drag the map. */
+  onDrag?: (center: LngLat, bounds: LngLatBounds) => void
+}
+
+/**
+ * A component that renders a map with markers to show result locations using Mapbox GL.
+ *
+ * @remarks
+ * For the map to work properly, be sure to include Mapbox GL stylesheet in the application.
+ *
+ * @param props - {@link MapboxMapProps}
+ * @returns A React element conatining a Mapbox Map
+ *
+ * @public
+ */
+export function MapboxMap<T>({
+  mapboxAccessToken,
+  mapboxOptions,
+  PinComponent,
+  getCoordinate = getDefaultCoordinate,
+  onDrag
+}: MapboxMapProps<T>) {
+  useEffect(() => {
+    mapboxgl.accessToken = mapboxAccessToken;
+  }, [mapboxAccessToken]);
+
+  const mapContainer = useRef<HTMLDivElement>(null);
+  const map = useRef<Map | null>(null);
+  const markers = useRef<Marker[]>([]);
+
+  const locationResults = useSearchState(state => state.vertical.results) as Result<T>[];
+  const hasLocationResults = locationResults && locationResults.length !== 0;
+  const onDragDebounced = useDebouncedFunction(onDrag, 100);
+
+  useEffect(() => {
+    if (mapContainer.current && !map.current) {
+      const options: MapboxOptions = {
+        container: mapContainer.current,
+        style: 'mapbox://styles/mapbox/streets-v11',
+        center: [-73.935242, 40.730610],
+        zoom: 9,
+        ...mapboxOptions
+      };
+      map.current = new Map(options);
+      const mapbox = map.current;
+      mapbox.resize();
+      if (onDragDebounced) {
+        mapbox.on('drag', () => {
+          onDragDebounced(mapbox.getCenter(), mapbox.getBounds());
+        });
+      }
+    }
+  }, [mapboxOptions, onDragDebounced]);
+
+  useEffect(() => {
+    markers.current.forEach(marker => marker.remove());
+    markers.current = [];
+
+    if (map.current && hasLocationResults) {
+      const bounds = new LngLatBounds();
+
+      for (const [index, result] of locationResults.entries()) {
```

Comment:
can this use declarative forEach instead?

Reply:
typescript fail to recognize the check if(map.current && ...) when using `forEach` -- complains about `map.current` being possibly null. This works fine for a normal for loop. There is workaround for it to use forEach so I can update to that